Implement workspace routing in singlehost mode

metlos opened the pull request:
### What does this PR do?
`$TITLE`

### What issues does this PR fix or reference?
eclipse/che#18583

metlos committed:
Singlehost routing for devworkspaces.

metlos committed:
Fix the URL format so that it is what che-theia expects, honor endpoint
protocol settings.

amisevsk commented on pkg/defaults/defaults.go:
```diff
@@ -14,12 +14,22 @@ const (
 	log = ctrl.Log.WithName("defaults")
 )
 
+func GetGatewayWorkpaceConfigMapName(workspaceID string) string {
+	return workspaceID
+}
```
We may want to prefix this with `che.routing.controller.devfile.io` -- if the main controller creates a configmap using workspaceID it will cause a conflict.

amisevsk replied:
If I understand correctly, this configmap is created in the controller's namespace, so feel free to ignore this comment.

amisevsk commented on pkg/solver/singlehost.go:
```diff
@@ -0,0 +1,325 @@
+
+			publicURLPrefix := getPublicURLPrefixForEndpoint(workspaceID, machineName, endpoint)
+
+			publicURL := scheme + "://" + ensureDoesntEndWithSlash(host) + ensureDoesntEndWithSlash(publicURLPrefix) + ensureStartsWithSlash(endpoint.Path)
```
Looks like a job for [`path.Join()`](https://golang.org/pkg/path/#Join) :)

metlos replied:
Oh, nice :)

amisevsk commented on pkg/solver/singlehost.go:
```diff
@@ -0,0 +1,325 @@
+}
+
+func (c *CheRoutingSolver) getGatewayConfigMaps(cheManager *dwoche.CheManager, workspaceID string, routing *dwo.WorkspaceRouting) []corev1.ConfigMap {
+	restrictedAnno, setRestrictedAnno := routing.Annotations[config.WorkspaceRestrictedAccessAnnotation]
+
+	labels := defaults.GetLabelsForComponent(cheManager, "gateway-config")
+	labels[config.WorkspaceIDLabel] = workspaceID
+	if setRestrictedAnno {
+		labels[config.WorkspaceRestrictedAccessAnnotation] = restrictedAnno
+	}
```
Careful when handling the restricted-access annotation. At the moment, che-machine-exec doesn't work if restricted access is set as we restrict pod/exec to the workspace owner.

Unless there's a specific usecase in mind, I would consider failing reconcile when restricted-access is set.

metlos replied:
I was trying to do the same thing that is being done in the `workspacerouting_controller.go` to services/ingresses/routes without actually understanding if this has any consequence since this is applied to a configmap in a different namespace from the namespace of the workspace.

amisevsk replied:
For configmaps it'll have no effect, since our webhooks don't listen for those (yet...). The main use cases are [don't allow pods/exec](https://github.com/devfile/devworkspace-operator/blob/c4ad8ce4d2ab25346daecab9ea1db3e933cbe081/webhook/workspace/handler/exec.go#L26-L53) and [don't allow objects to be modified](https://github.com/devfile/devworkspace-operator/blob/master/webhook/workspace/handler/immutable.go). Only the creator and the DevWorkspaceOperator can modify objects with `restricted-access` set. The reason we do this is specifically for the Web Terminal case, as someone could e.g. edit a service to intercept the call that injects a user's token.

metlos replied:
Ok, I'll leave the annotation in then, because basically the config maps are "restricted-access". Any change to them will be reverted by the operator to conform to what's in the routing.

amisevsk commented on pkg/solver/singlehost.go:
```diff
@@ -0,0 +1,325 @@
+  routers:` + routers + `
+  services:` + services + `
+  middlewares:` + middlewares + `
+`
```
Consider defining a struct to represent this and marshalling it via `sigs.k8s.io/yaml`, or alternatively using a nightmare Sprintf (you can use e.g. 
```golang
fmt.Sprintf(`%[1]s, %[1]s, %[1]s`, name)
```
to reuse arguments)

metlos replied:
Marshalling ftw.. Thanks for the suggestion!

amisevsk commented on pkg/solver/singlehost.go:
```diff
@@ -0,0 +1,325 @@
+				var prefix string
+				var serviceURL string
+
+				if endpointName == "" {
+					name = fmt.Sprintf("%s-%s-%d", workspaceID, machineName, port)
+				} else {
+					name = fmt.Sprintf("%s-%s-%d-%s", workspaceID, machineName, port, endpointName)
+				}
```
Watch out for the 63 char limit here.

metlos replied:
This is the contents of the traefik config file. I don't think traefik names have the 63 chars limit but actually this is an interesting thing to verify...

amisevsk replied:
Ah sorry, I misunderstood.

amisevsk commented on pkg/solver/solver.go:
```diff
@@ -42,42 +60,149 @@ func Getter(scheme *runtime.Scheme) *CheRouterGetter {
+	managers := manager.GetCurrentManagers()
+	if len(managers) == 0 {
+		// the CheManager has not been reconciled yet, so let's wait a bit
+		return &v1alpha1.CheManager{}, &solvers.RoutingNotReady{}
```
You can use the `Retry` param to delay the next reconcile:
```suggestion
		return &v1alpha1.CheManager{}, &solvers.RoutingNotReady{Retry: 1 * time.Second}
```

amisevsk reviewed (approved):
Didn't get a chance to test, but generally LGTM

metlos committed:
Explicitly set the retry delay when routing deployment is being delayed because of no che manager in the cluster.

Co-authored-by: Angel Misevski <amisevsk@redhat.com>

metlos committed:
Replace string concatenation with marshalling for greatly improved
safety and developer sanity.

metlos committed:
Add the missing import

metlos committed:
Use path.Join() instead of homegrown path composition functions.

sparkoo commented on pkg/solver/solver.go:
```diff
@@ -13,11 +13,26 @@
 package solver
```
What is the responsibility of `solver` package and this `solver.go` ?

metlos replied:
Added a `doc.go` in the package with a short description of what it's for.

sparkoo commented on pkg/manager/chemanager_controller.go:
```diff
@@ -31,21 +32,36 @@ import (
+	syncer  datasync.Syncer
+}
+
+func GetCurrentManagers() map[client.ObjectKey]v1alpha1.CheManager {
```
can you please explain to me why we need multiple managers and why we need to do it like this with mutex?

metlos replied:
Added the docs about what this is for and its pros and cons.

sparkoo commented on pkg/manager/chemanager_controller.go:
```diff
@@ -89,7 +105,17 @@ func (r *CheReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+		managerAccess.Lock()
+		defer managerAccess.Unlock()
+
+		currentManagers[req.NamespacedName] = *current
```
when this procedure happens (I mean whole `Reconcile` method)? Why we now need t remember the manager?

metlos replied:
The newly added comments on the `GetCurrentManagers()` method should explain this.

metlos committed:
Add some tests

sparkoo reviewed (commented):
I would very much welcome method docs and better var names. Also logic around multiple managers is not clear to me. I know what the code is doing, but I don't know why. It's probably just my lack of knowledge of operator-sdk.

metlos committed:
Docs on the GetCurrentManagers()

Retry routing reconciliation if we can't find the required che manager in
the cluster.

metlos committed:
Small udpates to the docs.

sparkoo commented on pkg/solver/doc.go:
```diff
@@ -0,0 +1,5 @@
+// Package solver contains the implementation of the "workspace routing solver" which provides che-specific
```
if this is only for "workspace routing solver", shouldn't it be in it's name, rather than generic `solver`? When this operator will have more responsibilities and potentially more "solvers", it will be hard to find.
If it won't be hard to rename later when we know better what we need, I'm ok keeping it as is for now.
